File: backend/src/main/java/br/ufpr/sept/so2/modules/iam/infrastructure/security/NimbusJwtTokenService.java
package br.ufpr.sept.so2.modules.iam.infrastructure.security;

import br.ufpr.sept.so2.modules.iam.application.ports.JwtTokenService;
import br.ufpr.sept.so2.modules.iam.domain.Usuario;
import br.ufpr.sept.so2.modules.iam.infrastructure.IamProperties;
import com.nimbusds.jose.JOSEObjectType;
import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.crypto.RSASSASigner;
import com.nimbusds.jose.crypto.RSASSAVerifier;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.security.KeyPair;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.time.Instant;
import java.util.Date;
import java.util.List;
import java.util.UUID;

@Component
public class NimbusJwtTokenService implements JwtTokenService {

    private static final Logger LOG = LoggerFactory.getLogger(NimbusJwtTokenService.class);

    private final IamProperties properties;
    private final RSAPrivateKey privateKey;
    private final RSAPublicKey publicKey;

    public NimbusJwtTokenService(IamProperties properties) {
        this.properties = properties;
        KeyPair pair = resolveKeyPair(properties);
        this.privateKey = (RSAPrivateKey) pair.getPrivate();
        this.publicKey = (RSAPublicKey) pair.getPublic();
    }

    @Override
    public String emitAccessToken(Usuario usuario) {
        Instant agora = Instant.now();
        JWTClaimsSet claims = new JWTClaimsSet.Builder()
                .subject(usuario.getId().toString())
                .issuer(properties.getIssuer())
                .audience(properties.getAudience())
                .jwtID(UUID.randomUUID().toString())
                .issueTime(Date.from(agora))
                .expirationTime(Date.from(agora.plusSeconds(properties.accessTtlSeconds())))
                .claim("authorities", usuario.getAuthorities())
                .claim("mustChangePassword", usuario.precisaPrimeiroAcesso())
                .build();
        return sign(claims);
    }

    @Override
    public String emitResetToken(Usuario usuario) {
        Instant agora = Instant.now();
        JWTClaimsSet claims = new JWTClaimsSet.Builder()
                .subject(usuario.getId().toString())
                .issuer(properties.getIssuer())
                .audience(properties.getResetAudience())
                .jwtID(UUID.randomUUID().toString())
                .issueTime(Date.from(agora))
                .expirationTime(Date.from(agora.plusSeconds(properties.resetTtlSeconds())))
                .build();
        return sign(claims);
    }

    @Override
    public AccessTokenClaims parseAccessToken(String token) {
        JWTClaimsSet claims = parse(token, properties.getAudience());
        List<String> authorities = readAuthorities(claims);
        boolean mustChange = Boolean.TRUE.equals(claims.getClaim("mustChangePassword"));
        return new AccessTokenClaims(
                UUID.fromString(claims.getSubject()),
                authorities,
                mustChange,
                claims.getJWTID()
        );
    }

    @Override
    public ResetTokenClaims parseResetToken(String token) {
        JWTClaimsSet claims = parse(token, properties.getResetAudience());
        return new ResetTokenClaims(UUID.fromString(claims.getSubject()), claims.getJWTID());
    }

    private String sign(JWTClaimsSet claims) {
        try {
            JWSHeader header = new JWSHeader.Builder(JWSAlgorithm.RS256)
                    .type(JOSEObjectType.JWT)
                    .keyID("so2-iam")
                    .build();
            SignedJWT jwt = new SignedJWT(header, claims);
            jwt.sign(new RSASSASigner(privateKey));
            return jwt.serialize();
        } catch (Exception ex) {
            throw new IllegalStateException("Falha ao assinar JWT", ex);
        }
    }

    private JWTClaimsSet parse(String token, String audienceEsperada) {
        try {
            SignedJWT jwt = SignedJWT.parse(token);
            if (!jwt.verify(new RSASSAVerifier(publicKey))) {
                throw new IllegalArgumentException("assinatura");
            }
            JWTClaimsSet claims = jwt.getJWTClaimsSet();
            Date exp = claims.getExpirationTime();
            if (exp == null || exp.before(new Date())) {
                throw new IllegalArgumentException("expirado");
            }
            if (claims.getAudience() == null || !claims.getAudience().contains(audienceEsperada)) {
                throw new IllegalArgumentException("audience");
            }
            if (claims.getSubject() == null || claims.getJWTID() == null) {
                throw new IllegalArgumentException("claims");
            }
            return claims;
        } catch (IllegalArgumentException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new IllegalArgumentException("JWT inválido", ex);
        }
    }

    @SuppressWarnings("unchecked")
    private static List<String> readAuthorities(JWTClaimsSet claims) {
        Object raw = claims.getClaim("authorities");
        if (raw instanceof List<?> list) {
            return list.stream().map(String::valueOf).toList();
        }
        return List.of();
    }

    private static KeyPair resolveKeyPair(IamProperties properties) {
        String privatePem = properties.getJwtPrivateKeyPem();
        String publicPem = properties.getJwtPublicKeyPem();
        if (privatePem != null && !privatePem.isBlank() && publicPem != null && !publicPem.isBlank()) {
            return new KeyPair(PemKeys.parsePublic(publicPem), PemKeys.parsePrivate(privatePem));
        }
        LOG.warn("Par RSA JWT ausente nas variáveis de ambiente; gerando chave efêmera (dev/test).");
        return PemKeys.generateRsa2048();
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/iam/infrastructure/security/NimbusJwtTokenService.kt
package br.ufpr.sept.so2.modules.iam.infrastructure.security

import br.ufpr.sept.so2.modules.iam.application.ports.JwtTokenService
import br.ufpr.sept.so2.modules.iam.domain.Usuario
import br.ufpr.sept.so2.modules.iam.infrastructure.IamProperties
import com.nimbusds.jose.JOSEObjectType
import com.nimbusds.jose.JWSAlgorithm
import com.nimbusds.jose.JWSHeader
import com.nimbusds.jose.crypto.RSASSASigner
import com.nimbusds.jose.crypto.RSASSAVerifier
import com.nimbusds.jwt.JWTClaimsSet
import com.nimbusds.jwt.SignedJWT
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.security.interfaces.RSAPrivateKey
import java.security.interfaces.RSAPublicKey
import java.time.Instant
import java.util.Date
import java.util.UUID

@Component
class NimbusJwtTokenService(
    private val properties: IamProperties,
) : JwtTokenService {
    private val privateKey: RSAPrivateKey
    private val publicKey: RSAPublicKey

    init {
        val pair = resolveKeyPair(properties)
        privateKey = pair.private as RSAPrivateKey
        publicKey = pair.public as RSAPublicKey
    }

    override fun emitAccessToken(usuario: Usuario): String {
        val agora = Instant.now()
        val claims = JWTClaimsSet.Builder()
            .subject(usuario.id.toString())
            .issuer(properties.issuer)
            .audience(properties.audience)
            .jwtID(UUID.randomUUID().toString())
            .issueTime(Date.from(agora))
            .expirationTime(Date.from(agora.plusSeconds(properties.accessTtlSeconds)))
            .claim("authorities", usuario.authorities)
            .claim("mustChangePassword", usuario.precisaPrimeiroAcesso())
            .build()
        return sign(claims)
    }

    override fun emitResetToken(usuario: Usuario): String {
        val agora = Instant.now()
        val claims = JWTClaimsSet.Builder()
            .subject(usuario.id.toString())
            .issuer(properties.issuer)
            .audience(properties.resetAudience)
            .jwtID(UUID.randomUUID().toString())
            .issueTime(Date.from(agora))
            .expirationTime(Date.from(agora.plusSeconds(properties.resetTtlSeconds)))
            .build()
        return sign(claims)
    }

    override fun parseAccessToken(token: String): JwtTokenService.AccessTokenClaims {
        val claims = parse(token, properties.audience)
        val authorities = readAuthorities(claims)
        val mustChange = claims.getClaim("mustChangePassword") == true
        return JwtTokenService.AccessTokenClaims(
            UUID.fromString(claims.subject),
            authorities,
            mustChange,
            claims.jwtid,
        )
    }

    override fun parseResetToken(token: String): JwtTokenService.ResetTokenClaims {
        val claims = parse(token, properties.resetAudience)
        return JwtTokenService.ResetTokenClaims(UUID.fromString(claims.subject), claims.jwtid)
    }

    private fun sign(claims: JWTClaimsSet): String {
        return try {
            val header = JWSHeader.Builder(JWSAlgorithm.RS256)
                .type(JOSEObjectType.JWT)
                .keyID("so2-iam")
                .build()
            val jwt = SignedJWT(header, claims)
            jwt.sign(RSASSASigner(privateKey))
            jwt.serialize()
        } catch (ex: Exception) {
            throw IllegalStateException("Falha ao assinar JWT", ex)
        }
    }

    private fun parse(token: String, audienceEsperada: String): JWTClaimsSet {
        try {
            val jwt = SignedJWT.parse(token)
            if (!jwt.verify(RSASSAVerifier(publicKey))) {
                throw IllegalArgumentException("assinatura")
            }
            val claims = jwt.jwtClaimsSet
            val exp = claims.expirationTime
            if (exp == null || exp.before(Date())) {
                throw IllegalArgumentException("expirado")
            }
            if (claims.audience == null || !claims.audience.contains(audienceEsperada)) {
                throw IllegalArgumentException("audience")
            }
            if (claims.subject == null || claims.jwtid == null) {
                throw IllegalArgumentException("claims")
            }
            return claims
        } catch (ex: IllegalArgumentException) {
            throw ex
        } catch (ex: Exception) {
            throw IllegalArgumentException("JWT inválido", ex)
        }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(NimbusJwtTokenService::class.java)

        private fun readAuthorities(claims: JWTClaimsSet): List<String> {
            val raw = claims.getClaim("authorities")
            if (raw is List<*>) {
                return raw.map { it.toString() }
            }
            return emptyList()
        }

        private fun resolveKeyPair(properties: IamProperties): java.security.KeyPair {
            val privatePem = properties.jwtPrivateKeyPem
            val publicPem = properties.jwtPublicKeyPem
            if (privatePem.isNotBlank() && publicPem.isNotBlank()) {
                return java.security.KeyPair(PemKeys.parsePublic(publicPem), PemKeys.parsePrivate(privatePem))
            }
            LOG.warn("Par RSA JWT ausente nas variáveis de ambiente; gerando chave efêmera (dev/test).")
            return PemKeys.generateRsa2048()
        }
    }
}
